### PR TITLE
Basic Authentication

### PR DESCRIPTION
**Pull request by pedro-cf**
**Related Issue(s):**

- #https://github.com/Healy-Hyperspatial/stac-fastapi-mongo/issues/12

**Description:**
Optional Basic Authentication for [stac-fastapi-mongo](https://github.com/Healy-Hyperspatial/stac-fastapi-mongo)
- I don't know how to create a stac_fastapi extension so I wrote a module `basic_auth.py` and added 1 function call to `app.py` `apply_basic_auth(api)`
- Added `docker-compose.basic_auth.yml` as an example on how to enable & configure basic auth:

```yml
- BASIC_AUTH_USER=admin
- BASIC_AUTH_PASS=admin
- BASIC_AUTH_ENDPOINTS=[{"path":"/collections","method":["POST","PUT","DELETE"]},{"path":"/collections/{collection_id}/items/{item_id}","method":["POST","PUT","DELETE"]}]
```
In order to enable basic auth both `BASIC_AUTH_USER` and `BASIC_AUTH_PASS` environment variables need to be set.

`BASIC_AUTH_ENDPOINTS` is optional and allows you to specify which paths/methods require authentication. If omitted all endpoints will require authentication.

After some additional review I will update further documention and tests if needed.

**PR Checklist:**

- [x] Code is formatted and linted (run `pre-commit run --all-files`)
- [x] Tests pass (run `make test`)
- [x] Documentation has been updated to reflect changes, if applicable
- [x] Changes are added to the changelog

### CODE REVIEW

**Comment by pedro-cf** [comment]
typing_extensions==4.4.0 seems to be missing from the checks? even do I added it to `setup.py`

**Comment by pedro-cf** [comment]
Could you take a look sir @jonhealy1 :) ?

**Comment by jonhealy1** [comment]
I have looked a little. It looks really good. Later today I will try it out. Like you said we will need a couple tests. It is definitely what I was thinking.

**Comment by jonhealy1** [comment]
If we add the logic [here](https://github.com/stac-utils/stac-fastapi-elasticsearch-opensearch/tree/main/stac_fastapi/core) to the core library then we  should be able to use this in the mongo, elasticsearch, and opensearch backends which would be really cool.

**Comment by jonhealy1** [comment]
We can add here too and then move it over later as well

File: docker-compose.basic_auth.yml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      - ENVIRONMENT=local
+      - BACKEND=mongo
+      - MONGO_DB=stac
+      - MONGO_HOST=mongo
+      - MONGO_USER=root
+      - MONGO_PASS=example
+      - MONGO_PORT=27017
+      - BASIC_AUTH_USER=admin
+      - BASIC_AUTH_PASS=admin
+      - BASIC_AUTH_ENDPOINTS=[{"path":"/collections","method":["POST","PUT","DELETE"]},{"path":"/collections/{collection_id}/items/{item_id}","method":["POST","PUT","DELETE"]}]
```
Comment (jonhealy1) [inline docker-compose.basic_auth.yml]: Being able to specify the endpoints here is really nice

**Comment by jonhealy1** [comment]
It's working great - just tested it

**Comment by pedro-cf** [comment]
@jonhealy1 Few Questions:

1. Would you prefer me to pull request to the elasticsearch repo or do both ?
2. Is this the right way to implement this? Would a Stacfastapi style extension be better (I don't know how) ? The same way these extensions are activated:
https://github.com/Healy-Hyperspatial/stac-fastapi-mongo/blob/ad755fde8b78cd17b2504361a4169e9003683109/stac_fastapi/mongo/app.py#L39-L59
https://github.com/Healy-Hyperspatial/stac-fastapi-mongo/blob/ad755fde8b78cd17b2504361a4169e9003683109/stac_fastapi/mongo/app.py#L65


4. Would you make any additions? f.e.:
 -  adding a `BASIC_AUTH_ENABLED`
 -  adding ability for multiple users with specific permissions (json: 
 ```json
 [{"username":"user","password":"pass","permissions": [{"path":"/collections","method":["POST","PUT","DELETE"]},{"path":"/collections/{collection_id}/items/{item_id}","method":["POST","PUT","DELETE"]}]}]
 ```
 -  adding ability for specifying which endpoints are public

**Comment by jonhealy1** [comment]
1. I think it may be easiest to add it to this repo first. Then add a pr to the other repo afterwards. It's really up to you though. 
 
- To add it to stac-fastapi-elasticsearch-opensearch and this project too we would add basic_auth.py to the [core](https://github.com/stac-utils/stac-fastapi-elasticsearch-opensearch/tree/main/stac_fastapi/core/stac_fastapi/core) library.
- Then we would push a new version of the core library to PyPI
- Update the core library version in stac-fastapi.elasticsearch,, stac-fastapi.opensearch, and stac-fastapi.mongo
- Then we could add the apply_basic_auth to the api of each
- Also, add the tests and add the docker compose file to each - elasticsearch and opensearch backends share the same tests while mongo has its own copy

**Comment by jonhealy1** [comment]
2. A lot of this depends on the sort of time you have right now. ie. If you just want to add to this repo and have me add it to the stac-fastapi-elasticsearch-opensearch repo. Adding an official extension would probably mean doing something here: https://stac-api-extensions.github.io/. We could define our own custom extension maybe. It could make sense to turn this into an 'official' extension later - I don't know. There's an example of an extension being defined [here](https://github.com/stac-utils/stac-fastapi/blob/main/stac_fastapi/extensions/stac_fastapi/extensions/third_party/bulk_transactions.py) that may help show how to turn this into a custom extension for now. There's a folder in the stac-fastapi-elasticsearch core library to hold extensions here: https://github.com/stac-utils/stac-fastapi-elasticsearch-opensearch/tree/main/stac_fastapi/core/stac_fastapi/core/extensions. If you want to tackle it go for it - make a custom extension for now.

**Comment by jonhealy1** [comment]
3. I think what you have now is good. It's nice and simple. I'm sort of 50/50 myself. I think most people just want a way to lock down the transactions endpoints but I guess also some people want to lock down all of the endpoints. Maybe an admin could make transactions but other registered Users could only view the data. What about one ENV variable only that sets everything like you showed above with users and routes? What do you think? You probably know as much or more than I do about what makes the most sense.

Would we need to specify which endpoints are public? Instead of listing which ones are prohibited we list the ones that are allowed? If this gets more complicated we will definitely need to add a section to the README

**Comment by pedro-cf** [comment]
@jonhealy1 

1. Alright

2. A custom extension seems like the correct option but I don't understand how to create it. I basically just need `apply_basic_auth(api)` to be called somewhere with the `api` as an argument.

3. I really think the approach with multiple users being defined in a json array is better, because it allows for all the api possibilities. Example:
   a. API where the [default] end points are public but [Transactions Extension] is protected. 
   b. API where the everything is protected. Would require 2 or more users (1 admin, and N readers)
Also by default endpoints without and permission associated could be set as public. A single env var would be enough for all of this.

**Comment by jonhealy1** [comment]
We could add like it is and then create an issue to convert it to a custom extension later? I agree with 3. let's do it, it is better.

**Comment by pedro-cf** [comment]
Greetings I've made some improvements:
- Single `BASIC_AUTH` JSON string Environment Variable: Combined all authentication settings into one variable
- The `public_endpoints` key allows to specfiy which endpoints are public
- You can set the key `permissions` as "*" to grant the user all permissions
- Provided some docker-compose examples for two use cases:

1. `docker-compose.basic_auth_public.yml`: all GET requests and `/search` POST are public. Other requests require admin authentication
```json
{
	"public_endpoints": [
		{"path": "/","method": "GET"},
		{"path": "/conformance","method": "GET"},
		{"path": "/collections/{collection_id}/items/{item_id}","method": "GET"},
		{"path": "/search","method": "GET"},
		{"path": "/search","method": "POST"},
		{"path": "/collections","method": "GET"},
		{"path": "/collections/{collection_id}","method": "GET"},
		{"path": "/collections/{collection_id}/items","method": "GET"},
		{"path": "/queryables","method": "GET"},
		{"path": "/queryables/collections/{collection_id}/queryables","method": "GET"},
		{"path": "/_mgmt/ping","method": "GET"}
	],
	"users": [
	  {
		"username": "admin",
		"password": "admin",
		"permissions": [
		  {"path": "/","method": ["GET"]},
		  {"path": "/conformance","method": ["GET"]},
		  {"path": "/collections/{collection_id}/items/{item_id}","method": ["GET","POST","PUT","DELETE"]},
		  {"path": "/search","method": ["GET","POST"]},
		  {"path": "/collections","method": ["GET","PUT","POST"]},
		  {"path": "/collections/{collection_id}","method": ["GET","DELETE"]},
		  {"path": "/collections/{collection_id}/items","method": ["GET","POST"]},
		  {"path": "/queryables","method": ["GET"]},
		  {"path": "/queryables/collections/{collection_id}/queryables","method": ["GET"]},
		  {"path": "/_mgmt/ping","method": ["GET"]}
		]
	  }
	]
}
```

2. `docker-compose.basic_auth_protected.yml`: All endpoints require authentication. The Admin user as access to all endpoints. The Reader user has access to all GET and `/search` POST endpoints
```json
{
	"users": [
	  {
		"username": "admin",
		"password": "admin",
		"permissions": "*"
	  },
	  {
		"username": "reader",
		"password": "reader",
		"permissions": [
		  {"path": "/","method": ["GET"]},
		  {"path": "/conformance","method": ["GET"]},
		  {"path": "/collections/{collection_id}/items/{item_id}","method": ["GET"]},
		  {"path": "/search","method": ["GET", "POST"]},
		  {"path": "/collections","method": ["GET"]},
		  {"path": "/collections/{collection_id}","method": ["GET"]},
		  {"path": "/collections/{collection_id}/items","method": ["GET"]},
		  {"path": "/queryables","method": ["GET"]},
		  {"path": "/queryables/collections/{collection_id}/queryables","method": ["GET"]},
		  {"path": "/_mgmt/ping","method": ["GET"]}
		]
	  }
	]
}
```

Side note, adding auth can be as simple as adding:
```yaml
- BASIC_AUTH={"users":[{"username":"user","password":"pass","permissions":"*"}]}
```

**Comment by pedro-cf** [comment]
@jonhealy1 

1. Would you make any changes?
2. I'm guessing there's some documentation to be done? Not sure where to add it.
3. Also some tests to be created? Not sure where or how.

**Comment by jonhealy1** [comment]
I think this looks really good. I will have a closer look tomorrow. For documentation we can add another section in the Readme. I am not entirely sure what tests will look like. We can add a new test file in the tests folder. If you don't have any tests by tomorrow I will look at what they could look like. Nice work. It's really good so far.

**Comment by pedro-cf** [comment]
@jonhealy1 There's one change I could see make sense....

Using a config file instead of an env var and have the env var be the file path... it would make this a bit more secure.

I had this branch where the passwords (even though in memory) are encrypted with `bcrypt` and the config file could be deleted after startup.

https://github.com/pedro-cf/stac-fastapi-mongo/blob/basic-auth-secure-ish/stac_fastapi/mongo/basic_auth.py

Would this be an ok addition? or do u think the env var is easier to use (no volumes etc)

**Comment by jonhealy1** [comment]
Definitely makes sense

**Comment by pedro-cf** [comment]
> Definitely makes sense

I actually changed my mind. I think this adds alot of deployment complexity and security isn't really a priority when using basic authentication, not to mention adding a file via-volume would have a file with plain text credentials in the machine anyway. So I think we should just abandon the idea.

Although timing attacks protection seems ok to add. 

Now I think we should just focus on docs/tests

**Comment by pedro-cf** [comment]
Regarding tests I'm assuming best way would be:

1. edit https://github.com/Healy-Hyperspatial/stac-fastapi-mongo/blob/ad755fde8b78cd17b2504361a4169e9003683109/.github/workflows/cicd.yml

2. add 1 or 2 steps/environments with BASIC_AUTH for each use case?

3. Exclude new tests from original test step, and only run specific tests for the new basic auth use cases steps ?

```yml
 - name: Run test suite against Mongo 
         run: | 
           pipenv run pytest -svvv -k "not basic_auth" ## Exclude basic_auth tests
         env: 
           MONGO_HOST: 172.17.0.1 
           BACKEND: mongo 
           APP_HOST: 0.0.0.0 
           APP_PORT: 8084 
           ENVIRONMENT: testing 
           MONGO_DB: stac 
           MONGO_USER: root 

 - name: Run test suite against basic auth public Mongo 
         run: | 
           pipenv run pytest ./tests/basic_auth/public_tests.py -svvv
         env: 
           MONGO_HOST: 172.17.0.1 
           BACKEND: mongo 
           APP_HOST: 0.0.0.0 
           APP_PORT: 8084 
           ENVIRONMENT: testing 
           MONGO_DB: stac 
           MONGO_USER: root 
           BASIC_AUTH: {"public_endpoints":[{"path":"/","method":"GET"},{"path":"/conformance","method":"GET"},{"path":"/collections/{collection_id}/items/{item_id}","method":"GET"},{"path":"/search","method":"GET"},{"path":"/search","method":"POST"},{"path":"/collections","method":"GET"},{"path":"/collections/{collection_id}","method":"GET"},{"path":"/collections/{collection_id}/items","method":"GET"},{"path":"/queryables","method":"GET"},{"path":"/queryables/collections/{collection_id}/queryables","method":"GET"},{"path":"/_mgmt/ping","method":"GET"}],"users":[{"username":"admin","password":"admin","permissions":[{"path":"/","method":["GET"]},{"path":"/conformance","method":["GET"]},{"path":"/collections/{collection_id}/items/{item_id}","method":["GET","POST","PUT","DELETE"]},{"path":"/search","method":["GET","POST"]},{"path":"/collections","method":["GET","PUT","POST"]},{"path":"/collections/{collection_id}","method":["GET","DELETE"]},{"path":"/collections/{collection_id}/items","method":["GET","POST"]},{"path":"/queryables","method":["GET"]},{"path":"/queryables/collections/{collection_id}/queryables","method":["GET"]},{"path":"/_mgmt/ping","method":["GET"]}]}]}

 - name: Run test suite against basic auth protected Mongo 
         run: | 
           pipenv run pytest ./tests/basic_auth/protected_tests.py -svvv
         env: 
           MONGO_HOST: 172.17.0.1 
           BACKEND: mongo 
           APP_HOST: 0.0.0.0 
           APP_PORT: 8084 
           ENVIRONMENT: testing 
           MONGO_DB: stac 
           MONGO_USER: root
           BASIC_AUTH: {"users":[{"username":"admin","password":"admin","permissions":"*"},{"username":"reader","password":"reader","permissions":[{"path":"/","method":["GET"]},{"path":"/conformance","method":["GET"]},{"path":"/collections/{collection_id}/items/{item_id}","method":["GET"]},{"path":"/search","method":["GET","POST"]},{"path":"/collections","method":["GET"]},{"path":"/collections/{collection_id}","method":["GET"]},{"path":"/collections/{collection_id}/items","method":["GET"]},{"path":"/queryables","method":["GET"]},{"path":"/queryables/collections/{collection_id}/queryables","method":["GET"]},{"path":"/_mgmt/ping","method":["GET"]}]}]}

```

What do you think @jonhealy1 ?

**Comment by jonhealy1** [comment]
Your idea for the tests is really good :) I think you make a good point about adding too much complexity as well.

**Comment by jonhealy1** [comment]
The docker compose files are a really good idea for showing functionality. I don't think we need too much for the docs - it's up to you

File: stac_fastapi/mongo/basic_auth.py [hunk 2]
```diff
@@ -72,17 +47,12 @@ def has_access(
...
-    ).digest()
-
-    # Compare the digests in constant time
-    if not constant_time_compare(actual_digest, expected_digest):
+    # Compare the provided username and password with the correct ones using compare_digest
+    if not secrets.compare_digest(
+        credentials.username.encode("utf-8"), user.get("username").encode("utf-8")
+    ) or not secrets.compare_digest(
+        credentials.password.encode("utf-8"), user.get("password").encode("utf-8")
+    ):
```
Comment (jonhealy1) [inline stac_fastapi/mongo/basic_auth.py:55]: Nice

**Comment by pedro-cf** [comment]
@jonhealy1 I've made some changes to `conftest.py`, added some tests and made changes the `cicd.yml`

Not sure where you would add some docs? section on `README.md` ?

**Comment by jonhealy1** [comment]
Tests are looking good. I think a section in the README is good for docs.

File: stac_fastapi/tests/basic_auth/test_basic_auth.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+import pytest
+
+# - BASIC_AUTH={"public_endpoints":[{"path":"/","method":"GET"},{"path":"/search","method":"GET"}],"users":[{"username":"admin","password":"admin","permissions":"*"},{"username":"reader","password":"reader","permissions":[{"path":"/conformance","method":["GET"]},{"path":"/collections/{collection_id}/items/{item_id}","method":["GET"]},{"path":"/search","method":["POST"]},{"path":"/collections","method":["GET"]},{"path":"/collections/{collection_id}","method":["GET"]},{"path":"/collections/{collection_id}/items","method":["GET"]},{"path":"/queryables","method":["GET"]},{"path":"/queryables/collections/{collection_id}/queryables","method":["GET"]},{"path":"/_mgmt/ping","method":["GET"]}]}]}
+
+
+@pytest.mark.asyncio
+async def test_get_search_not_authenticated(app_client_basic_auth):
+    """Test public endpoint search without auhtentication"""
```
Comment (jonhealy1) [inline stac_fastapi/tests/basic_auth/test_basic_auth.py]: Spelling - authentication :)

**Comment by jonhealy1** [comment]
So weird I don't know why the github workflow isn't running

**Comment by pedro-cf** [comment]
> So weird I don't know why the github workflow isn't running

Guessing its because i edited the cicd

**Comment by jonhealy1** [comment]
It's working now.

**Comment by pedro-cf** [comment]
@jonhealy1 I've added some documentation to the `README.md`, please take a look.

I'm not sure what else to change? `CHANGELOG.md` ?

**Review by jonhealy1 — APPROVED** [review]
Very good work :)

**Pull request merged**